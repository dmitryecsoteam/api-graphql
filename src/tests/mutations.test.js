const graphqlServer = require('../server/serverExpress');
const axios = require('axios');

const url = `http://localhost:${process.env.API_PORT}/${process.env.API_CONTEXT_ROOT}`;
const method = 'POST';
const headers = { 'Content-Type': 'application/json' };

jest.mock('../models/Destination.js');
jest.mock('../models/Origin.js');
jest.mock('../models/Travel.js');
jest.mock('../models/User.js');
jest.mock('../auth/JWT.js');

const save = require('../models/User').save;

let server;

beforeAll(async () => {
    server = await graphqlServer.init();
});

afterAll(async () => {
    await server.close();
    console.log("Server stopped");
});

describe('signinUser mutation', () => {
    test('should signin user and return token', async () => {

        const data = JSON.stringify({ query: 'mutation {signinUser(email: "test@mail.com", password: "XXX") { token }}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
    });

    test('should return error if user doesn\'t exist', async () => {

        const data = JSON.stringify({ query: 'mutation {signinUser(email: "randomUser@mail.com", password: "XXX") { token }}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
    });

    test('should return error if password doesn\'t match', async () => {

        const data = JSON.stringify({ query: 'mutation {signinUser(email: "test@mail.com", password: "YYY") { token }}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
    });
});

describe('signupUser mutation', () => {
    test('should sign up new user and return token', async () => {

        const data = JSON.stringify({ query: 'mutation {signupUser(email: "test1@mail.com", password: "XXX1", name: "User Name1") { token }}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
    });

    test('should not sign up existing user and return error', async () => {

        const data = JSON.stringify({ query: 'mutation {signupUser(email: "test@mail.com", password: "XXX1", name: "User Name1") { token }}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
    });
});

describe('addNotification mutation', () => {
    test('should add new notification by id to existing user', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(id: "5d239408f05e9620082a2419") { email notifications { travelId date priceAirplaneLast priceHotelLast } }}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(1);
    });

    test('should add new notification by date/origin/dest to existing user', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(date: "2019-07-07", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(1);
    });

    test('should not add notification by id if it already exists', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(id: "5c6c7e42e3b7e1edb974a40c", date: "2019-08-09", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast } }}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not add notification by date/origin/dest if it already exists', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(date: "2019-08-09", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not add notification if user doesn\'t exist', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(date: "2019-08-09", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' })

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer non_exist_user_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not add notification by id if id doesn\'t exist', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(id: "5c6c7e42e3b7e1edb974a40d", date: "2019-08-09", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast } }}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not add notification by date/origin/dest if date is past or out of range', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(date: "2000-01-10", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not add notification by date/origin/dest if origin or destination doesn\'t exist', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(date: "2019-07-08", origin: 100, destination: 200) { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not add notification without authorization token', async () => {

        const data = JSON.stringify({ query: 'mutation {addNotification(id: "5d239408f05e9620082a2419", date: "2019-07-07", origin: 1, destination: 2) { email notifications { travelId date priceAirplaneLast priceHotelLast } }}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

});

describe('deleteNotification mutation', () => {
    test('should delete notification', async () => {

        const data = JSON.stringify({ query: 'mutation {deleteNotification(id: "5c6c7e42e3b7e1edb974a40c") { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });
        
        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(1);
    });

    test('should return error if user doesn\'t exist', async () => {

        const data = JSON.stringify({ query: 'mutation {deleteNotification(id: "5c6c7e42e3b7e1edb974a40c") { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer non_exist_user_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should return error if notification doesn\'t exist', async () => {

        const data = JSON.stringify({ query: 'mutation {deleteNotification(id: "5c6c7e42e3b7e1edb974a40d") { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers: { ...headers, Authorization: 'Bearer valid_token' }, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });

    test('should not delete notification without authorization token', async () => {

        const data = JSON.stringify({ query: 'mutation {deleteNotification(id: "5c6c7e42e3b7e1edb974a40c") { email notifications { travelId date priceAirplaneLast priceHotelLast }}}' });

        const response = await axios({ method, url, headers, data });

        expect(response.data).toMatchSnapshot();
        expect(save).toHaveBeenCalledTimes(0);
    });
});