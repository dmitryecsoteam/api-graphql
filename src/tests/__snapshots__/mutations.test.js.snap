// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addNotification mutation should add new notification by date/origin/dest to existing user 1`] = `
Object {
  "data": Object {
    "addNotification": Object {
      "email": "test@mail.com",
      "notifications": Array [
        Object {
          "date": "2019-08-09",
          "priceAirplaneLast": null,
          "priceHotelLast": 75,
          "travelId": "5c6c7e42e3b7e1edb974a40c",
        },
        Object {
          "date": "2019-07-07",
          "priceAirplaneLast": 53,
          "priceHotelLast": null,
          "travelId": "5bbce95f81e97acfcd8e0a76",
        },
      ],
    },
  },
}
`;

exports[`addNotification mutation should add new notification by id to existing user 1`] = `
Object {
  "data": Object {
    "addNotification": Object {
      "email": "test@mail.com",
      "notifications": Array [
        Object {
          "date": "2019-08-09",
          "priceAirplaneLast": null,
          "priceHotelLast": 75,
          "travelId": "5c6c7e42e3b7e1edb974a40c",
        },
        Object {
          "date": "2019-07-07",
          "priceAirplaneLast": null,
          "priceHotelLast": 25,
          "travelId": "5d239408f05e9620082a2419",
        },
      ],
    },
  },
}
`;

exports[`addNotification mutation should not add notification by date/origin/dest if it already exists 1`] = `
Object {
  "data": Object {
    "addNotification": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "Such notification already exists. TravelId: 5c6c7e42e3b7e1edb974a40c, date: 2019-08-09",
      "path": Array [
        "addNotification",
      ],
    },
  ],
}
`;

exports[`addNotification mutation should not add notification by id if it already exists 1`] = `
Object {
  "data": Object {
    "addNotification": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "Such notification already exists. TravelId: 5c6c7e42e3b7e1edb974a40c, date: 2019-08-09",
      "path": Array [
        "addNotification",
      ],
    },
  ],
}
`;

exports[`addNotification mutation should not add notification if user doesn't exist 1`] = `
Object {
  "data": Object {
    "addNotification": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "User with email random_user@mail.com doesn't exist",
      "path": Array [
        "addNotification",
      ],
    },
  ],
}
`;

exports[`deleteNotification mutation should delete notification 1`] = `
Object {
  "data": Object {
    "deleteNotification": Object {
      "email": "test@mail.com",
      "notifications": Array [],
    },
  },
}
`;

exports[`deleteNotification mutation should return error if notification doesn't exist 1`] = `
Object {
  "data": Object {
    "deleteNotification": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "User test@mail.com doesn't have notification with id 5c6c7e42e3b7e1edb974a40d",
      "path": Array [
        "deleteNotification",
      ],
    },
  ],
}
`;

exports[`deleteNotification mutation should return error if user doesn't exist 1`] = `
Object {
  "data": Object {
    "deleteNotification": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "User with email random_user@mail.com doesn't exist",
      "path": Array [
        "deleteNotification",
      ],
    },
  ],
}
`;

exports[`signinUser mutation should return error if password doesn't match 1`] = `
Object {
  "data": Object {
    "signinUser": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "Wrong password",
      "path": Array [
        "signinUser",
      ],
    },
  ],
}
`;

exports[`signinUser mutation should return error if user doesn't exist 1`] = `
Object {
  "data": Object {
    "signinUser": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "User with email randomUser@mail.com doesn't exist",
      "path": Array [
        "signinUser",
      ],
    },
  ],
}
`;

exports[`signinUser mutation should signin user and return token 1`] = `
Object {
  "data": Object {
    "signinUser": Object {
      "token": "test_token",
    },
  },
}
`;

exports[`signupUser mutation should not sign up existing user and return error 1`] = `
Object {
  "data": Object {
    "signupUser": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 11,
          "line": 1,
        },
      ],
      "message": "User with email test@mail.com already exists",
      "path": Array [
        "signupUser",
      ],
    },
  ],
}
`;

exports[`signupUser mutation should sign up new user and return token 1`] = `
Object {
  "data": Object {
    "signupUser": Object {
      "token": "test_token",
    },
  },
}
`;
